Manage "Expect" header with cURL transport

carlalexander opened the pull request:
Fixes #453

carlalexander commented:
I'm not sure how you prefer to handle the `strlen` calculation since `data` can be an array or a string. I did it inside the `strlen` happy to change it to something else.

jrfnl commented:
@carlalexander Thanks for finding this and for the PR!

Do you think this could be covered by unit tests in some way ?

> I'm not sure how you prefer to handle the strlen calculation since data can be an array or a string. I did it inside the strlen happy to change it to something else.

For code readability and comprehension, I'd much prefer it if it would be pulled apart and calculated before the `strlen()`, but that's just me.

Also, for the multi-line comment, you may want to use multi-comment style `/*... */` ;-)

carlalexander commented:
Normally, I do multi-line comments that way too. I'm just trying to keep the same coding style as the one below. Happy to switch it to the `/*`.

I can also create a function to get the get the `strlen`.

For testing, I'd like to add tests. I just didn't really see anything for testing cURL code. The cURL transport has 4 tests to check for some thrown exceptions. That's it.

jrfnl commented:
> For testing, I'd like to add tests. I just didn't really see anything for testing cURL code.

I see where your confusion comes from. Most of the tests for the cURL code are in the `RequestsTest_Transport_Base` test class which the `RequestsTest_Transport_cURL` test class extends.
The `RequestsTest_Transport_cURL` class only contains the cURL specific tests, so I imagine a test for this cURL specific issue would be best placed in that file.

If you run the tests with `phpunit --testdox --filter RequestsTest_Transport_cURL`, you can see that the cURL class is tested with a lot more tests than just those in the one file.

jrfnl commented:
> Normally, I do multi-line comments that way too. I'm just trying to keep the same coding style as the one below. Happy to switch it to the /*.

I hear you. We've cleaned up the code style recently, cleaning up the comments is on the to do list still, ;-)

carlalexander commented:
Alright, I'll take a look at adding tests over the weekend. 

I've done another pass and fixed up the comment and moved the `strlen` code to its own method. I added the method at the end because I can't seem to figure out how methods are ordered. Just let me know where you'd like it.

carlalexander commented:
Alright, so I've tried to write some tests, but the Heroku application that we run tests against doesn't seem to return the `Expect` header back. I've tried with both an empty header and one with `100-Continue`. 

Is the code for that Heroku application anywhere? Also it could be that Heroku strips the header.

jrfnl commented:
@carlalexander I'm not sure at all, but I suspect it might be this repo: https://github.com/rmccue/httpbin

@rmccue Ryan, could you please confirm ? (or point to the real source)

carlalexander commented:
Yeah, I'm trying against `https://httpbin.org/headers` and it's not returning the header regardless that its set or not.

schlessera commented:
I think the source of the server is this one: https://github.com/RequestsPHP/test-server

rmccue commented:
> I think the source of the server is this one: https://github.com/RequestsPHP/test-server

That's the one, yep :)

jrfnl commented:
@carlalexander Just checking in - will you have time to update the PR for the tests ? And if so, what timeline are we looking at ?

carlalexander commented:
@jrfnl Hoping to sit down and hammer at it some more this weekend. 😄  I haven't looked at it this week.

carlalexander commented:
Ok, so I have some tests running, but they only work with @rmccue's server. I'm just running it locally atm. Are there tests that use it? It seems everything uses httpbin.org and it doesn't return the header. 

Another good news is you can clearly see the 1 sec timeout with the tests. So that's great.

<img width="541" alt="image" src="https://user-images.githubusercontent.com/654684/111881316-45868980-8986-11eb-9edd-81da6f754a0a.png">

jrfnl commented:
@carlalexander I hope I understood your question correctly, but AFAICS, all tests which use a function call to `httpbin()` to set the address use the custom test server.

Also see #315 for some additional context.

carlalexander commented:
Ok, this helped @jrfnl! I tried with `requests-php-tests.herokuapp.com` and it also strips the header. But Travis seems to run the tests against the test server running with the built-in PHP server. So I pushed the tests up and they seem to work on Travis!

jrfnl commented:
@carlalexander Awesome! I'll have a more detailed look soon. What about annotating the tests with the performance fix with [`@small`](https://phpunit.readthedocs.io/en/9.5/annotations.html#small) to ensure they fail if the extra second would be added ?

jrfnl commented:
@carlalexander Was just looking at Travis and I notice there does still seem to be an issue:
```
1) RequestsTest_Transport_cURL::testGETWithNestedData

Array to string conversion

/home/travis/build/WordPress/Requests/library/Requests/Transport/cURL.php:573
/home/travis/build/WordPress/Requests/library/Requests/Transport/cURL.php:332
/home/travis/build/WordPress/Requests/library/Requests/Transport/cURL.php:135
/home/travis/build/WordPress/Requests/library/Requests.php:381
/home/travis/build/WordPress/Requests/tests/Transport/Base.php:111

2) RequestsTest_Transport_cURL::testPOSTWithNestedData

Array to string conversion

/home/travis/build/WordPress/Requests/library/Requests/Transport/cURL.php:573
/home/travis/build/WordPress/Requests/library/Requests/Transport/cURL.php:332
/home/travis/build/WordPress/Requests/library/Requests/Transport/cURL.php:135
/home/travis/build/WordPress/Requests/library/Requests.php:381
/home/travis/build/WordPress/Requests/library/Requests.php:270
/home/travis/build/WordPress/Requests/tests/Transport/Base.php:222
```

carlalexander commented:
@jrfnl That's my bad. I was using the new function signature for `implode`. Fixed it. Also added `@small` tags.

carlalexander commented:
Ok, well the old signature makes `phpcs` fail now lol

jrfnl commented:
> using the new function signature for implode.

There is no "new" function signature. There is a cross-version compatible signature and a second signature which was previously supported, but isn't anymore.

The function should be called like so: `implode ( string $separator , array $array )`

> Ok, well the old signature makes phpcs fail now lol

And rightfully so as using the reverse order will throw a deprecation notice on PHP 7.4 (and fail the tests because of it) and be a fatal error on PHP 8.0.

carlalexander commented:
Yeah, I'm reverting the change. I understand what's going on now. I'm also going to add some more tests.

jrfnl commented:
@carlalexander Appreciated! You're a ⭐

carlalexander commented:
Everything is green except for PHP 8.0. It doesn't seem to be an issue with the tests or code. It seems to be PHPUnit related.

jrfnl commented:
> Everything is green except for PHP 8.0. It doesn't seem to be an issue with the tests or code. It seems to be PHPUnit related.

Correct - see #439

carlalexander commented:
Perfect, should be good to merge once you've done your review. Let me know if you need anything else 😄

schlessera commented on library/Requests/Transport/cURL.php:
```diff
@@ -546,4 +561,23 @@ public static function test($capabilities = array()) {
+	 * @return string The "Expect" header
+	 */
+	protected function get_expect_header($data) {
+		if (is_array($data)) {
+			$iterator = new RecursiveIteratorIterator(new RecursiveArrayIterator($data));
+			$data     = '';
+
+			foreach ($iterator as $datum) {
+				$data .= $datum;
+			}
+		}
+
+		return strlen($data) >= 1048576 ? '100-Continue' : '';
```
This is not very efficient given we're only interested in the byte size of the array. The concatenation of the string can be expensive in terms of memory operations, and then you still haven't counted the bytes yet.

How about this instead:
```suggestion
		if (is_array($data)) {
			$iterator = new RecursiveIteratorIterator(new RecursiveArrayIterator($data));
			$bytesize = 0;

			foreach ($iterator as $datum) {
				$bytesize += strlen((string)$datum);
			}
		} else {
			$bytesize = strlen((string)$data);
		}

		return $bytesize >= 1048576 ? '100-Continue' : '';
```

carlalexander commented:
Pushed @schlessera's recommended changes 😄

TimothyBJacobs commented on library/Requests/Transport/cURL.php:
```diff
@@ -546,4 +561,25 @@ public static function test($capabilities = array()) {
+			$bytesize = 0;
+
+			foreach ($iterator as $datum) {
+				$bytesize += strlen((string) $datum);
```
I think it might be worth bailing out here immediately once the `bytesize` limit is reached.

carlalexander commented:
Really great suggestion @TimothyBJacobs! I've reworked the function to early return if we hit the 1MB limit.

schlessera commented:
Dammit! I broke the PR when I wanted to do a rebase. When will I learn not to touch PRs that were done directly in `master`?

@carlalexander Are you able to push the change again rebased on latest `master`? If not, let me know, and I'll recreate your PR.

schlessera commented:
I recreated the PR here: #469
